File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

import Prelude hiding (lookup)

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

earthYearLength :: Float;
earthYearLength = 31557600;

yearLength:: Planet -> Float
yearLength Mercury= 0.2408467;
yearLength Venus= 0.61519726;
yearLength Earth= 1.0;
yearLength Mars= 1.8808158;
yearLength Jupiter= 11.862615;
yearLength Saturn= 29.447498;
yearLength Uranus= 84.016846;
yearLength Neptune= 164.79132;

ageOn :: Planet -> Float -> Float
ageOn Earth seconds = seconds / earthYearLength;
ageOn planet seconds = (1 / yearLength planet) * ageOn Earth seconds

File: typescript/space-age/space-age.ts
type Planet = 'mercury' | 'venus' | 'earth' | 'mars' | 'jupiter' | 'saturn' | 'uranus' | 'neptune';

const earhYearInSeconds =  31557600;

function yearLength(planet: Planet): number {
  switch (planet) {
    case 'mercury': return 0.2408467;
    case 'venus': return 0.61519726;
    case 'earth': return 1.0;
    case 'mars': return 1.8808158;
    case 'jupiter': return 11.862615;
    case 'saturn': return 29.447498;
    case 'uranus': return 84.016846;
    case 'neptune': return 164.79132;
  };
}


console.log(yearLength);


export function age(planet: Planet, seconds: number): number {
  const age = (seconds / earhYearInSeconds) * ( 1 / yearLength(planet));
  return Math.round(age * 100) / 100;
}
